cmd/synchronization-controller/synchronization-controller.go

```diff
@@ -0,0 +1,426 @@
+
+	ctx context.Context
+}
+
+func (app *synchronizationControllerApp) prepareCertManager() (err error) {
+	app.clientcertmanager = bootstrap.NewFileCertificateManager(app.clientCertFilePath, app.clientKeyFilePath)
+	app.servercertmanager = bootstrap.NewFileCertificateManager(app.serverCertFilePath, app.serverKeyFilePath)
+	return
+}
+
+// Update synchronization controller log verbosity on relevant config changes
+func (app *synchronizationControllerApp) shouldChangeLogVerbosity() {
+	verbosity := app.clusterConfig.GetVirtSynchronizationControllerVerbosity()
+	if verbosity == 0 {
+		// If the verbosity gets set in kubevirt CR, this will not be 0, but it is otherwise.
+		verbosity = 2
+	}
+	err := log.Log.SetVerbosityLevel(int(verbosity))
+	if err != nil {
+		log.Log.Errorf("unable to change log verbosity to %d, %v", verbosity, err)
+	} else {
+		log.Log.V(2).Infof("set verbosity to %d", verbosity)
+	}
+}
+
+// Update synchronization controller rate limiter
+func (app *synchronizationControllerApp) shouldChangeRateLimiter() {
+	config := app.clusterConfig.GetConfig()
+	qps := config.HandlerConfiguration.RestClient.RateLimiter.TokenBucketRateLimiter.QPS
+	burst := config.HandlerConfiguration.RestClient.RateLimiter.TokenBucketRateLimiter.Burst
+	app.reloadableRateLimiter.Set(flowcontrol.NewTokenBucketRateLimiter(qps, burst))
+	log.Log.V(2).Infof("setting rate limiter to %v QPS and %v Burst", qps, burst)
+}
+
+func (app *synchronizationControllerApp) setupTLS(factory controller.KubeInformerFactory) error {
+	kubevirtCAConfigInformer := factory.KubeVirtCAConfigMap()
+	if err := kubevirtCAConfigInformer.SetWatchErrorHandler(func(r *cache.Reflector, err error) {
+		apiHealthVersion.Clear()
+		cache.DefaultWatchErrorHandler(r, err)
+	}); err != nil {
+		return err
+	}
+
+	app.caManager = kvtls.NewCAManager(kubevirtCAConfigInformer.GetStore(), app.namespace, app.caConfigMapName)
+
+	app.serverTLSConfig = kvtls.SetupTLSForVirtSynchronizationControllerServer(app.caManager, app.servercertmanager, app.externallyManaged, app.clusterConfig)
+	app.clientTLSConfig = kvtls.SetupTLSForVirtSynchronizationControllerClients(app.caManager, app.clientcertmanager, app.externallyManaged)
+	return nil
+}
+
+func (app *synchronizationControllerApp) Run() {
+	var err error
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	app.ctx = ctx
+
+	app.LeaderElection = leaderelectionconfig.DefaultLeaderElectionConfiguration()
+
+	app.reloadableRateLimiter = ratelimiter.NewReloadableRateLimiter(flowcontrol.NewTokenBucketRateLimiter(virtconfig.DefaultVirtControllerQPS, virtconfig.DefaultVirtHandlerBurst))
+	var clientConfig *rest.Config
+	for i := 0; i < maxRetryCount; i++ {
+		clientConfig, err = kubecli.GetKubevirtClientConfig()
+		if err != nil {
+			log.Log.Errorf("unable to get kubevirt client config %v", err)
+			waitTime := 2 ^ (i + 1)
+			time.Sleep(time.Duration(waitTime) * time.Millisecond)
+			continue
+		}
+		i = maxRetryCount
+	}
```

Comment:
Should we maybe exit/panic now if `err != nil`? (same below)

Reply:
No the point is that maybe some things might not be ready (like api server) and doing some retries will help us avoid some crash loops. But I did just notice I am not checking if the maxRetry is reached, at which point we should panic.

Reply:
That's what I was trying to say :)